File: src/rules/StorageNode.cpp
#include <QtCore/QString>

#include "StorageNode.h"
#include "../Brain.h"

namespace CnotiMind
{

	StorageNode::StorageNode(const QString& key, const QString& value, MemoryType memory, Brain* brain, QObject* parent):
		RuleNode( key, value, brain, parent ),
		_memory( memory )
	{

	}

	/*
		Creates a memory Event and store it in the brain memory
	*/
	void StorageNode::exec()
	{
		MemoryEvent m( _key, _value );

		_brain->storeToMemory( m, _memory);
	}

	/*

	*/
	void StorageNode::exec( QHash<QString, QString> &variables )
	{
		MemoryEvent m( _key, variableToValue( _value, variables ) );

		_brain->storeToMemory( m, _memory);
	}

	QString StorageNode::info( int depth ) const
	{
		QString info;

		info += space(depth) + "Storage key=" + _key + " value=" + _value;

		return info + RuleNode::info(depth);
	}
}

File: src_ios/CnotiMind/StorageNode.m
//
//  StorageNode.m
//  CnotiMind
//
//  Created by Gonçalo Rodrigues on 11/05/02.
//  Copyright 2011 Universidade de Coimbra. All rights reserved.
//

#import "StorageNode.h"
#import "MemoryEvent.h"
#import "Brain.h"


@implementation StorageNode


- (id) initWithEventAndValueAndAndBrainAndParent: (NSString*)aEvent 
                                           value:(NSString*)aValue 
                                          memory:(enum MemoryType)aMemory
                                           brain:(Brain*)aBrain 
                                          parent:(id)aParent
{
    if (self == [super initWithBrainAndParent:aBrain parent:aParent]) {
        
        _memory = aMemory;
        _event = aEvent;
        _value = aValue;
        _clearStorage = FALSE;
    }

    return self;
}

/*
 Creates a memory Event and store it in the brain memory
 */
- (void) exec
{
    // Test if the node is for clearing the memory
    if( _clearStorage )
    {
        if( _memory == WorkingMemory )
        {
            [_brain clearWorkingMemory];
        }
        else
        {
            [_brain clearMemory];
        }
        return;
    }
    // If it is not for clearing, check if the value and event are valid
    if( [_event length]==0 || [_value length]==0 )
    {
        return;
    }
    
    // Create memory event to add to the memory
    MemoryEvent* m = [[MemoryEvent alloc] initWithEventAndValue:_event value:_value];
    [_brain storeToMemory:m memoryType:_memory];

}


- (void) exec:(NSMutableDictionary*)aVariables
{
//    MemoryEvent* m = [[MemoryEvent alloc] initWithEventAndValue:_event value: [self variableToValue:_value variables:aVariables]];
    [self exec];
}


- (NSString*) info:(int)aDepth
{
    NSString* info = [NSString stringWithFormat:@""];
    NSString* space = [self space:aDepth];
    
    info = [info stringByAppendingFormat:@"%@ Storage key=%@ value=%@", space, _event, _value];
    
    return info;
}



@end
